File: core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/pipelines/MaxSpeedEnvelope.java
package fr.sncf.osrd.envelope_sim.pipelines;

import static fr.sncf.osrd.envelope.part.constraints.EnvelopePartConstraintType.CEILING;
import static fr.sncf.osrd.envelope.part.constraints.EnvelopePartConstraintType.FLOOR;
import static fr.sncf.osrd.envelope_sim.TrainPhysicsIntegrator.POSITION_EPSILON;

import fr.sncf.osrd.envelope.Envelope;
import fr.sncf.osrd.envelope.EnvelopeCursor;
import fr.sncf.osrd.envelope.OverlayEnvelopeBuilder;
import fr.sncf.osrd.envelope.part.ConstrainedEnvelopePartBuilder;
import fr.sncf.osrd.envelope.part.EnvelopePartBuilder;
import fr.sncf.osrd.envelope.part.constraints.EnvelopeConstraint;
import fr.sncf.osrd.envelope.part.constraints.SpeedConstraint;
import fr.sncf.osrd.envelope_sim.EnvelopeProfile;
import fr.sncf.osrd.envelope_sim.EnvelopeSimContext;
import fr.sncf.osrd.envelope_sim.StopMeta;
import fr.sncf.osrd.envelope_sim.overlays.EnvelopeDeceleration;
import fr.sncf.osrd.reporting.exceptions.OSRDError;

/** Max speed envelope = MRSP + braking curves
 * It is the max speed allowed at any given point, ignoring allowances
 */
public class MaxSpeedEnvelope {
    static boolean increase(double prevPos, double prevSpeed, double nextPos, double nextSpeed) {
        // Works for both accelerations (forwards) and decelerations (backwards)
        return prevSpeed < nextSpeed;
    }

    /** Generate braking curves overlay everywhere the mrsp decrease (increase backwards) with a discontinuity */
    private static Envelope addBrakingCurves(EnvelopeSimContext context, Envelope mrsp) {
        var builder = OverlayEnvelopeBuilder.backward(mrsp);
        var cursor = EnvelopeCursor.backward(mrsp);
        var lastPosition = mrsp.getEndPos();
        while (cursor.findPartTransition(MaxSpeedEnvelope::increase)) {
            if (cursor.getPosition() > lastPosition) {
                // The next braking curve already covers this point, this braking curve is hidden
                cursor.nextPart();
                continue;
            }
            var partBuilder = new EnvelopePartBuilder();
            partBuilder.setAttr(EnvelopeProfile.BRAKING);
            var overlayBuilder = new ConstrainedEnvelopePartBuilder(
                    partBuilder,
                    new SpeedConstraint(0, FLOOR),
                    new EnvelopeConstraint(mrsp, CEILING)
            );
            var startSpeed = cursor.getSpeed();
            var startPosition = cursor.getPosition();
            // TODO: link directionSign to cursor boolean reverse
            EnvelopeDeceleration.decelerate(context, startPosition, startSpeed, overlayBuilder, -1);
            builder.addPart(partBuilder.build());
            cursor.nextPart();
            lastPosition = overlayBuilder.getLastPos();
        }
        return builder.build();
    }

    /** Generate braking curves overlay at every stop position */
    private static Envelope addStopBrakingCurves(
            EnvelopeSimContext context,
            double[] stopPositions,
            Envelope curveWithDecelerations
    ) {
        for (int i = 0; i < stopPositions.length; i++) {
            var stopPosition = stopPositions[i];
            // if the stopPosition is zero, no need to build a deceleration curve
            if (stopPosition == 0.0)
                continue;
            if (stopPosition > curveWithDecelerations.getEndPos()) {
                if (Math.abs(stopPosition - curveWithDecelerations.getEndPos()) < POSITION_EPSILON)
                    stopPosition = curveWithDecelerations.getEndPos();
                else
                    throw OSRDError.newEnvelopeError(i, stopPosition, curveWithDecelerations.getEndPos());
            }
            var partBuilder = new EnvelopePartBuilder();
            partBuilder.setAttr(EnvelopeProfile.BRAKING);
            partBuilder.setAttr(new StopMeta(i));
            var overlayBuilder = new ConstrainedEnvelopePartBuilder(
                    partBuilder,
                    new SpeedConstraint(0, FLOOR),
                    new EnvelopeConstraint(curveWithDecelerations, CEILING)
            );
            EnvelopeDeceleration.decelerate(context, stopPosition, 0, overlayBuilder, -1);

            var builder = OverlayEnvelopeBuilder.backward(curveWithDecelerations);
            builder.addPart(partBuilder.build());
            curveWithDecelerations = builder.build();
        }
        return curveWithDecelerations;
    }

    /** Generate a max speed envelope given a mrsp */
    public static Envelope from(EnvelopeSimContext context, double[] stopPositions, Envelope mrsp) {
        var maxSpeedEnvelope = addBrakingCurves(context, mrsp);
        maxSpeedEnvelope = addStopBrakingCurves(context, stopPositions, maxSpeedEnvelope);
        return maxSpeedEnvelope;
    }
}

File: core/envelope-sim/src/main/kotlin/fr/sncf/osrd/envelope_sim/pipelines/MaxSpeedEnvelope.kt
package fr.sncf.osrd.envelope_sim.pipelines

import fr.sncf.osrd.envelope.Envelope
import fr.sncf.osrd.envelope.EnvelopeCursor
import fr.sncf.osrd.envelope.OverlayEnvelopeBuilder
import fr.sncf.osrd.envelope.part.ConstrainedEnvelopePartBuilder
import fr.sncf.osrd.envelope.part.EnvelopePartBuilder
import fr.sncf.osrd.envelope.part.constraints.EnvelopeConstraint
import fr.sncf.osrd.envelope.part.constraints.EnvelopePartConstraintType
import fr.sncf.osrd.envelope.part.constraints.SpeedConstraint
import fr.sncf.osrd.envelope_sim.EnvelopeProfile
import fr.sncf.osrd.envelope_sim.EnvelopeSimContext
import fr.sncf.osrd.envelope_sim.etcs.ETCSBrakingSimulator
import fr.sncf.osrd.envelope_sim.etcs.ETCSBrakingSimulatorImpl
import fr.sncf.osrd.envelope_sim.etcs.EndOfAuthority
import fr.sncf.osrd.envelope_sim.etcs.EoaType
import fr.sncf.osrd.envelope_sim.overlays.EnvelopeDeceleration
import fr.sncf.osrd.path.interfaces.TravelledPath
import fr.sncf.osrd.railjson.schema.schedule.RJSTrainStop
import fr.sncf.osrd.utils.units.Distance
import fr.sncf.osrd.utils.units.Offset

/**
 * Max speed envelope = MRSP + braking curves It is the max speed allowed at any given point,
 * ignoring allowances
 */

/**
 * Simple data class used to pass stop data to the class. Combines the stop offset with the
 * closed/open signal flag.
 */
data class SimStop(
    val offset: Offset<TravelledPath>,
    val rjsReceptionSignal: RJSTrainStop.RJSReceptionSignal,
)

fun increase(prevPos: Double, prevSpeed: Double, nextPos: Double, nextSpeed: Double): Boolean {
    // Works for both accelerations (forwards) and decelerations (backwards)
    return prevSpeed < nextSpeed
}

/**
 * Generate braking curves overlay everywhere the mrsp decrease (increase backwards) with a
 * discontinuity
 */
private fun addSlowdownBrakingCurves(
    etcsSimulator: ETCSBrakingSimulator,
    context: EnvelopeSimContext,
    mrsp: Envelope,
): Envelope {
    var envelope = mrsp
    envelope = addETCSSlowdownBrakingCurves(etcsSimulator, envelope)
    envelope = addConstSlowdownBrakingCurves(context, envelope)
    return envelope
}

/**
 * Generate braking curves overlay everywhere the mrsp decreases (increases backwards) with a
 * discontinuity using constant deceleration (outside ETCS ranges).
 */
private fun addConstSlowdownBrakingCurves(
    context: EnvelopeSimContext,
    envelope: Envelope,
): Envelope {
    val builder = OverlayEnvelopeBuilder.backward(envelope)
    val cursor = EnvelopeCursor.backward(envelope)
    var lastPosition = envelope.endPos
    while (cursor.findPartTransition(::increase)) {
        if (cursor.position > lastPosition) {
            // The next braking curve already covers this point, this braking curve is hidden
            cursor.nextPart()
            continue
        }
        val partBuilder = EnvelopePartBuilder()
        partBuilder.setAttr(EnvelopeProfile.BRAKING)
        val overlayBuilder =
            ConstrainedEnvelopePartBuilder(
                partBuilder,
                SpeedConstraint(0.0, EnvelopePartConstraintType.FLOOR),
                EnvelopeConstraint(envelope, EnvelopePartConstraintType.CEILING),
            )
        val startSpeed = cursor.speed
        val startPosition = cursor.position
        // TODO: link directionSign to cursor boolean reverse
        EnvelopeDeceleration.decelerate(context, startPosition, startSpeed, overlayBuilder, -1.0)
        builder.addPart(partBuilder.build())
        cursor.nextPart()
        lastPosition = overlayBuilder.lastPos
    }
    return builder.build()
}

/** Add braking curves following ETCS rules in relevant places */
private fun addETCSSlowdownBrakingCurves(
    etcsSimulator: ETCSBrakingSimulator,
    envelope: Envelope,
): Envelope {
    val limitsOfAuthority = etcsSimulator.computeLoaLocations(envelope)
    return etcsSimulator.addSlowdownBrakingCurves(envelope, limitsOfAuthority)
}

/** Generate braking curves overlay at every stop position */
private fun addStopBrakingCurves(
    etcsSimulator: ETCSBrakingSimulator,
    context: EnvelopeSimContext,
    stops: List<SimStop>,
    curveWithDecelerations: Envelope,
): Envelope {
    var envelope = curveWithDecelerations
    val etcsStops =
        etcsSimulator.computeEoaLocations(
            curveWithDecelerations,
            stops.map { it.offset },
            stops.map { it.rjsReceptionSignal.isStopOnClosedSignal },
            EoaType.STOP,
        )
    val constStops =
        stops
            .filter { stop ->
                stop.offset.distance != Distance.ZERO &&
                    !etcsStops.map { it.offsetEOA }.contains(stop.offset)
            }
            .map { it.offset }
    envelope = addETCSStopBrakingCurves(etcsSimulator, envelope, etcsStops)
    envelope = addConstStopBrakingCurves(context, envelope, constStops)
    return envelope
}

/** Generate braking curves overlay at every stop position */
private fun addConstStopBrakingCurves(
    context: EnvelopeSimContext,
    curveWithDecelerations: Envelope,
    stops: List<Offset<TravelledPath>>,
): Envelope {
    var envelope = curveWithDecelerations
    for (stop in stops) {
        val partBuilder = EnvelopePartBuilder()
        partBuilder.setAttr(EnvelopeProfile.BRAKING)
        val overlayBuilder =
            ConstrainedEnvelopePartBuilder(
                partBuilder,
                SpeedConstraint(0.0, EnvelopePartConstraintType.FLOOR),
                EnvelopeConstraint(envelope, EnvelopePartConstraintType.CEILING),
            )
        EnvelopeDeceleration.decelerate(context, stop.meters, 0.0, overlayBuilder, -1.0)

        val builder = OverlayEnvelopeBuilder.backward(envelope)
        builder.addPart(partBuilder.build())
        envelope = builder.build()
    }
    return envelope
}

/** Add braking parts for any ETCS flagged stop. */
private fun addETCSStopBrakingCurves(
    simulator: ETCSBrakingSimulator,
    envelope: Envelope,
    stops: List<EndOfAuthority>,
): Envelope {
    return simulator.addStopBrakingCurves(envelope, stops)
}

/** Generate a max speed envelope given a mrsp */
fun maxSpeedEnvelopeFrom(
    context: EnvelopeSimContext,
    stops: List<SimStop>,
    mrsp: Envelope,
): Envelope {
    val etcsSimulator = ETCSBrakingSimulatorImpl(context)
    var maxSpeedEnvelope = addSlowdownBrakingCurves(etcsSimulator, context, mrsp)
    maxSpeedEnvelope = addStopBrakingCurves(etcsSimulator, context, stops, maxSpeedEnvelope)
    return maxSpeedEnvelope
}
